redun/executors/gcp_utils.py

```diff
@@ -224,3 +234,105 @@ def get_task(client: batch_v1.BatchServiceClient, task_name: str) -> batch_v1.Ta
         A Task object representing the specified task.
     """
     return client.get_task(name=task_name)
+
+
+# largely based on nextflow's implementation: https://github.com/nextflow-io/nextflow/blob/master/plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchMachineTypeSelector.groovy
+@dataclass
+class MachineType:
+    type: str
+    family: str
+    spot_price: float
+    on_demand_price: float
+    cpus: int
+    memory: int
+    gpus: int
+
+
+@cache
+def get_available_machine_types(region: str) -> List[MachineType]:
+    CLOUD_INFO_API = "https://cloudinfo.seqera.io/api/v1"
+    request = requests.get(
+        f"{CLOUD_INFO_API}/providers/google/services/compute/regions/{region}/products",
+        timeout=30,
+    )
+    if request.status_code != 200:
+        raise RuntimeError(f"Error getting machine types: {request.text}")
+    result: List[Dict] = request.json()["products"]
+    return [
+        MachineType(
+            type=p["type"],
+            family=p["type"].split("-")[0],
+            spot_price=mean(e["price"] for e in p["spotPrice"]),
+            on_demand_price=p["onDemandPrice"],
+            cpus=p["cpusPerVm"],
+            memory=p["memPerVm"],
+            gpus=p["gpusPerVm"],
+        )
+        for p in result
+    ]
+
+
+def find_best_matching_machine_type(
+    cpus: int,
+    memory: int,
+    region: str,
+    spot: bool,
+    local_ssd: bool,
+    families: Optional[List[str]] = None,
+    min_cpu_platform: Optional[MinCPUPlatform] = None,
+) -> str:
+    FAMILY_COST_CORRECTION = {
+        "e2": 1.0,  # Mix of processors, tend to be similar in performance to N1
+        # INTEL
+        "n1": 1.0,  # Skylake, Broadwell, Haswell, Sandy Bridge, and Ivy Bridge ~2.7 Ghz
+        "n2": 0.85,  # Intel Xeon Gold 6268CL ~3.4 Ghz
+        "c2": 0.8,  # Intel Xeon Gold 6253CL ~3.8 Ghz
+        "m1": 1.0,  # Intel Xeon E7-8880V4 ~2.7 Ghz
+        "m2": 0.85,  # Intel Xeon Platinum 8280L ~3.4 Ghz
+        "m3": 0.85,  # Intel Xeon Platinum 8373C ~3.4 Ghz
+        "a2": 0.9,  # Intel Xeon Platinum 8273CL ~2.9 Ghz
+        # AMD
+        "t2d": 1.0,  # AMD EPYC Milan ~2.7 Ghz
+        "n2d": 1.0,  # AMD EPYC Milan ~2.7 Ghz
+    }
+    FAMILIES_WITH_LOCAL_SSD = ["n1", "n2", "n2d", "c2", "c2d", "m3"]
+    NO_MIN_CPU_PLATFORM = ["e2-micro", "e2-small", "e2-medium", "f1-micro", "g1-small"]
+
+    machine_types = get_available_machine_types(region)
```

Comment:
if the sequera API fails, then this will throw an error too. we should try, catch this error and return the default to make sure the run doesn't fail in that case